64.CppCompat/main.cpp

```diff
@@ -36,10 +172,132 @@ struct T
+    }
+
+    void onAppTerminated_impl() override
+    {
+        m_logicalDevice->waitIdle();
+    }
+
+    void workLoopBody() override
+    {
+        waitForFrame(1u, m_fence);
+        m_logicalDevice->resetFences(1u, &m_fence.get());
+
+        m_cmdbuf->reset(video::IGPUCommandBuffer::ERF_RELEASE_RESOURCES_BIT);
+        m_cmdbuf->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+
+        m_cmdbuf->bindComputePipeline(m_pipeline.get());
+        m_cmdbuf->dispatch(1u, 1u, 1u);
+        m_cmdbuf->end();
+
+            //TODO: fix semaphores
+        CommonAPI::Submit(
+            m_logicalDevice.get(),
+            m_cmdbuf.get(),
+            m_queues[CommonAPI::InitOutput::EQT_COMPUTE],
+            m_semaphores[0].get(),
+            m_semaphores[1].get(),
+            m_fence.get());
+
+        m_keepRunning = false;
+    }
 
+    bool keepRunning() override
+    {
+        return m_keepRunning;
+    }
 
-int main()
+    static void runTests(int argc, char** argv)
+    {
+        CommonAPI::main<CompatibilityTest>(argc, argv);
+    }
+
+    void setWindow(core::smart_refctd_ptr<nbl::ui::IWindow>&& wnd) override {}
+    void setSystem(core::smart_refctd_ptr<nbl::system::ISystem>&& system) override { m_system = std::move(system); }
+    void setSurface(core::smart_refctd_ptr<video::ISurface>&& s) override {}
+    void setFBOs(std::vector<core::smart_refctd_ptr<video::IGPUFramebuffer>>& f) override {}
+    void setSwapchain(core::smart_refctd_ptr<video::ISwapchain>&& s) override {}
+    nbl::ui::IWindow* getWindow() override { return nullptr; }
+    video::IAPIConnection* getAPIConnection() override { return nullptr; }
+    video::ILogicalDevice* getLogicalDevice()  override { return m_logicalDevice.get(); }
+    video::IGPURenderpass* getRenderpass() override { return nullptr; }
+    uint32_t getSwapchainImageCount() override { return 0u; }
+    virtual nbl::asset::E_FORMAT getDepthFormat() override { return nbl::asset::E_FORMAT::EF_UNKNOWN;  }
+    APP_CONSTRUCTOR(CompatibilityTest);
+
+private:
+    core::smart_refctd_ptr<nbl::system::ISystem> m_system;
+    core::smart_refctd_ptr<nbl::video::ILogicalDevice> m_logicalDevice;
+    core::smart_refctd_ptr<video::IGPUComputePipeline> m_pipeline = nullptr;
+    core::smart_refctd_ptr<video::IGPUCommandBuffer> m_cmdbuf = nullptr;
+    std::array<video::IGPUQueue*, CommonAPI::InitOutput::MaxQueuesCount> m_queues;
+    core::smart_refctd_ptr<video::IGPUSemaphore> m_semaphores[2u] = {nullptr};
+    core::smart_refctd_ptr<video::IGPUFence> m_fence = nullptr;
+    std::array<std::array<nbl::core::smart_refctd_ptr<nbl::video::IGPUCommandPool>, CommonAPI::InitOutput::MaxFramesInFlight>, CommonAPI::InitOutput::MaxQueuesCount> commandPools;
+    core::smart_refctd_ptr<nbl::asset::IAssetManager> m_assetManager;
+    video::IGPUObjectFromAssetConverter::SParams m_cpu2gpuParams;
+    core::smart_refctd_ptr<nbl::system::ILogger> m_logger;
+
+    bool m_keepRunning = true;
+};
```

Comment:
noooo, this needs to be a 100% GPU-less test

Reply:
then how do we check if it works with hlsl?

Reply:
> then how do we check if it works with hlsl?

our cpp_compat should already give you 1:1 with HLSL when math is concerned

Reply:
it is usefull for me so will leave it for now